internal/ls/utilities.go

```diff
@@ -724,3 +810,114 @@ func getAllSuperTypeNodes(node *ast.Node) []*ast.TypeNode {
+	return start <= textRange.Pos() && end >= textRange.End()
+}
+
+func getPossibleGenericSignatures(called *ast.Expression, typeArgumentCount int, c *checker.Checker) []*checker.Signature {
+	typeAtLocation := c.GetTypeAtLocation(called)
+	if ast.IsOptionalChain(called.Parent) {
+		typeAtLocation = removeOptionality(typeAtLocation, ast.IsOptionalChainRoot(called.Parent), true /*isOptionalChain*/, c)
+	}
+	var signatures []*checker.Signature
+	if ast.IsNewExpression(called.Parent) {
+		signatures = c.GetSignaturesOfType(typeAtLocation, checker.SignatureKindConstruct)
+	} else {
+		signatures = c.GetSignaturesOfType(typeAtLocation, checker.SignatureKindCall)
+	}
+	return core.Filter(signatures, func(s *checker.Signature) bool {
+		return s.TypeParameters() != nil && len(s.TypeParameters()) >= typeArgumentCount
+	})
+}
+
+func removeOptionality(t *checker.Type, isOptionalExpression bool, isOptionalChain bool, c *checker.Checker) *checker.Type {
+	if isOptionalExpression {
+		return c.GetNonNullableType(t)
+	} else if isOptionalChain {
+		return c.GetNonOptionalType(t)
+	}
+	return t
+}
+
+func isNoSubstitutionTemplateLiteral(node *ast.Node) bool {
+	return node.Kind == ast.KindNoSubstitutionTemplateLiteral
+}
+
+func isTaggedTemplateExpression(node *ast.Node) bool {
+	return node.Kind == ast.KindTaggedTemplateExpression
+}
+
+func isInsideTemplateLiteral(node *ast.Node, position int, sourceFile *ast.SourceFile) bool {
+	return ast.IsTemplateLiteralKind(node.Kind) && (scanner.GetTokenPosOfNode(node, sourceFile, false) < position && position < node.End() || (ast.IsUnterminatedLiteral(node) && position == node.End()))
+}
+
+// Pseudo-literals
+func isTemplateHead(node *ast.Node) bool {
+	return node.Kind == ast.KindTemplateHead
+}
+
+func isTemplateTail(node *ast.Node) bool {
+	return node.Kind == ast.KindTemplateTail
+}
+
+func findPrecedingMatchingToken(token *ast.Node, matchingTokenKind ast.Kind, sourceFile *ast.SourceFile) *ast.Node {
+	closeTokenText := scanner.TokenToString(token.Kind)
+	matchingTokenText := scanner.TokenToString(matchingTokenKind)
+	tokenFullStart := token.Loc.Pos()
+	// Text-scan based fast path - can be bamboozled by comments and other trivia, but often provides
+	// a good, fast approximation without too much extra work in the cases where it fails.
+	bestGuessIndex := strings.LastIndex(sourceFile.Text(), matchingTokenText)
+	if bestGuessIndex == -1 {
+		return nil // if the token text doesn't appear in the file, there can't be a match - super fast bail
+	}
+	// we can only use the textual result directly if we didn't have to count any close tokens within the range
+	if strings.LastIndex(sourceFile.Text(), closeTokenText) < bestGuessIndex {
+		nodeAtGuess := astnav.FindPrecedingToken(sourceFile, bestGuessIndex+1)
+		if nodeAtGuess != nil && nodeAtGuess.Kind == matchingTokenKind {
+			return nodeAtGuess
+		}
+	}
+	tokenKind := token.Kind
+	remainingMatchingTokens := 0
+	for true {
```

Comment:
I think the test of the codebases uses 

```suggestion
	for {
```